server/api/crud/alerts.py

```diff
@@ -0,0 +1,260 @@
+
+                if len(state_obj["events"]) >= alert.criteria.count:
+                    send_notification = True
+            else:
+                send_notification = True
+
+            active = False
+            if send_notification:
+                state.count += 1
+                logger.debug("Sending notifications for alert", name=alert.name)
+                AlertNotificationPusher().push(alert, event_data)
+
+                if alert.reset_policy == "auto":
+                    self.reset_alert(session, alert.project, alert.name)
+                else:
+                    active = True
+
+            server.api.utils.singletons.db.get_db().store_alert_state(
+                session,
+                alert.project,
+                alert.name,
+                count=state.count,
+                last_updated=event_data.timestamp,
+                obj=state_obj,
+                active=active,
+            )
+
+    @staticmethod
+    def _validate_alert(alert, name, project):
+        if name != alert.name:
+            raise mlrun.errors.MLRunBadRequestError(
+                f"Alert name mismatch for alert {name} for project {project}. Provided {alert.name}"
+            )
+
+        if (
+            alert.criteria is not None
+            and alert.criteria.period is not None
+            and server.api.utils.helpers.string_to_timedelta(
+                alert.criteria.period, raise_on_error=False
+            )
+            is None
+        ):
+            raise mlrun.errors.MLRunBadRequestError(
+                f"Invalid period ({alert.criteria.period}) specified for alert {name} for project {project}"
+            )
+
+        for notification in alert.notifications:
+            if notification.kind not in [
+                mlrun.common.schemas.NotificationKind.git,
+                mlrun.common.schemas.NotificationKind.slack,
+                mlrun.common.schemas.NotificationKind.webhook,
+            ]:
+                raise mlrun.errors.MLRunBadRequestError(
+                    f"Unsupported notification ({notification.kind}) for alert {name} for project {project}"
+                )
+            notification_object = mlrun.model.Notification.from_dict(
+                notification.dict()
+            )
+            notification_object.validate_notification()
+
+        if alert.entity.project != project:
+            raise mlrun.errors.MLRunBadRequestError(
+                f"Invalid alert entity project ({alert.entity.project}) for alert {name} for project {project}"
+            )
+
+    @staticmethod
+    def _normalize_events(obj, period):
+        now = datetime.datetime.now(tz=datetime.timezone.utc)
+        events = obj["events"]
+        for event in events:
+            if isinstance(event, str):
+                event_time = datetime.datetime.fromisoformat(event)
+            else:
+                event_time = event
+            if event_time > now + period:
+                events.remove(event)
```

Comment:
This seems strange to me - aren't the events here the set of events kept in the alert state? How can these events be newer than `now + period`? Or are we expecting `period` to be negative?
Just looking at the code I would expect this to be `now > event_time + period` or something similar.
Please verify this, and if it isn't any of the reasons I mentioned, then please add a comment to explain the logic here since it is confusing.

Reply:
you are right. good catch